src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+	int i;
+	
+	prologue = prologue ? prologue : "";
+	prefix   = prefix   ? prefix   : "";
+	suffix   = suffix   ? suffix   : "";
+	epilogue = epilogue ? epilogue : "";
+	
+	strb_appends(s, prologue);
+	for(i=startIdx;i<endIdx;i++){
+		strb_appendf(s, "%s%d%s%s", prefix, i, suffix, &","[i==endIdx-1]);
+	}
+	strb_appends(s, epilogue);
+}
+static void  appendRangeCalculations(strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	int i;
+	
+	strb_appends(s, "\t/* Compute ranges for this thread. */\n");
+	
+	for(i=0;i<ctx->numIdx    ;i++){/* i*Dim   = srcSize[*]; */
+		strb_appendf(s, "\ti%dDim   = srcSize[%d];\n", i, i);
+	}
+	for(i=0;i<ctx->numIdx    ;i++){/* i*SStep = srcSteps[*]; */
+		strb_appendf(s, "\ti%dSStep = srcSteps[%d];\n", i, i);
+	}
+	for(i=0;i<ctx->numFreeIdx;i++){/* i*MStep = dstMaxSteps[*]; */
+		strb_appendf(s, "\ti%dMStep = dstMaxSteps[%d];\n", i, i);
+	}
+	for(i=0;i<ctx->numFreeIdx;i++){/* i*AStep = dstArgmaxSteps[*]; */
+		strb_appendf(s, "\ti%dMStep = dstArgmaxSteps[%d];\n", i, i);
+	}
+	for(i=0;i<ctx->numIdx    ;i++){/* i*Blk   = numBlk[*]; */
+		strb_appendf(s, "\ti%dBlk   = numBlk[%d];\n", i, i);
+	}
+	for(i=ctx->numIdx-1;i>=ctx->numFreeIdx;i--){/* i*PDim  = ...; */
+		/**
+		 * If this is the last index, it's the first cumulative dimension
+		 * product we generate, and thus we initialize to 1.
+		 */
+		
+		if(i == ctx->numIdx-1){
+			strb_appendf(s, "\ti%dPDim  = 1;\n", i);
+		}else{
+			strb_appendf(s, "\ti%dPDim  = i%dPDim * i%dDim;\n", i, i+1, i);
+		}
+	}
+	for(i=0;i<ctx->numIdx    ;i++){/* i*Start = ...; */
+		/**
+		 * The first 3 dimensions get to rely on hardware loops.
+		 * The others, if any, have to use software looping beginning at 0.
+		 */
+		
+		if(i < 3){
+			strb_appendf(s, "\ti%dStart = ((bi%d * bd%d) + ti%d) * i%dBlk;\n", i, i, i, i, i);
+		}else{
+			strb_appendf(s, "\ti%dStart = 0;\n", i);
+		}
+	}
+	for(i=0;i<ctx->numIdx    ;i++){/* i*End   = ...; */
+		/**
+		 * The first 3 dimensions get to rely on hardware loops.
+		 * The others, if any, have to use software looping beginning at 0.
+		 */
+		
+		if(i < 3){
+			strb_appendf(s, "\ti%dEnd   = i%dStart + bd%d * i%dBlk;\n", i, i, i, i);
+		}else{
+			strb_appendf(s, "\ti%dEnd   = i%dStart + i%dDim;\n", i, i, i);
+		}
+	}
+	
+	strb_appends(s, "\t\n");
+	strb_appends(s, "\t\n");
+}
+static void  appendLoops            (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	strb_appends(s, "\t/**\n");
+	strb_appends(s, "\t * FREE LOOPS.\n");
+	strb_appends(s, "\t */\n");
+	strb_appends(s, "\t\n");
+	
+	appendLoopMacroDefs  (s, ctx);
+	appendLoopOuter      (s, ctx);
+	appendLoopMacroUndefs(s, ctx);
+}
+static void  appendLoopMacroDefs    (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	int i;
+	
+	/**
+	 * FOROVER Macro
+	 */
+	
+	strb_appends(s, "#define FOROVER(idx)    for(i##idx = i##idx##Start; i##idx < i##idx##End; i##idx++)\n");
```

Comment:
@abergeron Defining the macro, once, the way I've done it allows me to greatly simplify the `strb_appendf()` that would otherwise result. Generating

    FOROVER(0){ESCAPE(0)
    FOROVER(1){ESCAPE(1)
    FOROVER(2){ESCAPE(2)
    FOROVER(3){ESCAPE(3)

is far easier, using `strb_appendf()`, than
    for(i0 = i0Start; i0<i0End; i0++)
    for(i1 = i1Start; i1<i1End; i1++)
    for(i2 = i2Start; i2<i2End; i2++)
    for(i3 = i3Start; i3<i3End; i3++)